servlet/src/main/java/io/undertow/servlet/spec/ServletOutputStreamImpl.java

```diff
@@ -871,20 +889,13 @@ public void handleEvent(final StreamSinkChannel aChannel) {
                 try {
                     state |= FLAG_IN_CALLBACK;
 
-                    servletRequestContext.getCurrentServletContext().invokeOnWritePossible(servletRequestContext.getExchange(), listener);
-
-                    if (isReady()) {
-                        //if the stream is still ready then we do not resume writes
-                        //this is per spec, we only call the listener once for each time
-                        //isReady returns true
-                        if (channel != null) {
-                            channel.suspendWrites();
-                        }
-                    } else {
-                        if (channel != null) {
-                            channel.resumeWrites();
-                        }
+                    //if the stream is still ready then we do not resume writes
+                    //this is per spec, we only call the listener once for each time
+                    //isReady returns true
+                    if (channel != null) {
+                        channel.suspendWrites();
```

Comment:
One potential issue with this is that is can lead to performance issues with NIO, due to the way it implements java.nio.channels.SelectionKey#interestOps(int) (which is what suspend/resume translates to internally). Basically each suspend/resume call is queued, and processed when select() is called. An approach like this that can result in many suspend/resumes between select() calls can result in a large number of changes being queued, which can adversely impact performance. This is not a huge hit, but it is not ideal either. 

On this flip side the alternative is to implement a lazy approach, which is potentially much more complex and may require stronger thread safety guarantees, which can also impact performance.

All things considered I am happy to go with your approach to start with, and if it results in performance issues then look at a more complex solution.